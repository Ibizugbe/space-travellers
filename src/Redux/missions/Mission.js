const JOIN_MISSION = 'space-travelers/mission/JOIN_MISSION';
const LEAVE_MISSION = 'space-travelers/mission/LEAVE_MISSION';
const FETCH_MISSION = 'space-travelers/mission/FETCH_MISSION';
const urlLink = 'https://api.spacexdata.com/v3/missions';

export const joinMission = (id) => ({
  type: JOIN_MISSION,
  payload: {
    id,
  },
});

export const fetchMission = (payload) => ({
  type: FETCH_MISSION,
  payload,
});

export const leaveMission = (id) => ({
  type: LEAVE_MISSION,
  payload: {
    id,
  },
});

export const fetchMissionsAPI = () => async (dispatch) => {
  const res = await fetch(urlLink);
  const mission = await res.json();
  const missionList = mission.map((mission) => ({
    id: mission.mission_id,
    name: mission.mission_name,
    description: mission.description,
  }));
  dispatch(fetchMission(missionList));
};

// reducer
const missionReducer = (state = [], action) => {
  switch (action.type) {
    case JOIN_MISSION: {
      const updateState = state.map((mission) => {
        if (mission.id !== action.payload.id) {
          return mission;
        }
        return { ...mission, reserved: true };
      });
      return [...updateState];
    }

    case LEAVE_MISSION: {
      const updateState = state.map((mission) => {
        if (mission.id !== action.payload.id) {
          return mission;
        }
        return { ...mission, reserved: false };
      });
      return [...updateState];
    }
    case FETCH_MISSION:
      return [...action.payload];
    default:
      return state;
  }
};

export default missionReducer;
