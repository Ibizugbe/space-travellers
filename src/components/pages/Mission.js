import React from 'react';
import { useSelector } from 'react-redux';
import DisplayMission from '../missions/DisplayMissions';

const Mission = () => {
  const missions = useSelector((state) => state.mission);
  console.log(missions);
  return (
    <table>
      <thead>
        <tr>
          <th>Missions</th>
          <th>Description</th>
          <th>Status</th>
        </tr>
      </thead>
      <tbody>
        {missions.map((mission) => (
          <DisplayMission
            name={mission.name}
            key={mission.id}
            id={mission.id}
            description={mission.description}
            reserved={mission.reserved}
          />
        ))}
      </tbody>
    </table>
  );
};

export default Mission;
