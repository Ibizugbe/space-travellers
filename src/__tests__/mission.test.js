import React from "react";
import Renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import store from "../redux/configureStore";
import Mission from "../components/pages/Mission";

it("renders correctly", () => {
  const tree = Renderer.create(
    <React.StrictMode>
      <Provider store={store}>
        <BrowserRouter>
          <Mission />
        </BrowserRouter>
      </Provider>
    </React.StrictMode>
  ).toJSON();
  expect(tree).toMatchSnapshot(`
<div
  className="container-fluid px-5"
>
  <table
    className="table table-hover table-striped table-responsive"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Missions
        </th>
        <th
          scope="col"
        >
          Description
        </th>
        <th
          scope="col"
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`);
});
