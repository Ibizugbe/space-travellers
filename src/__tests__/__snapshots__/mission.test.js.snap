// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly: 
<div
  className="container-fluid px-5"
>
  <table
    className="table table-hover table-striped table-responsive"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Missions
        </th>
        <th
          scope="col"
        >
          Description
        </th>
        <th
          scope="col"
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
 1`] = `
<div
  className="container-fluid px-5"
>
  <table
    className="table table-hover table-striped table-responsive"
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Missions
        </th>
        <th
          scope="col"
        >
          Description
        </th>
        <th
          scope="col"
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
